Feature/threading gui and pt

jellosubmarine opened the pull request.

Commit:
Added simple async

Commit:
Changed bind to lambda. Changed function order.

Commit:
Threading started working, picture broke.

Commit:
Fixed frame counting for PT

Commit:
Added non-working OMP and changed rendering loop

Commit:
Working OMP

Commit:
Added wait to resetBuffer for sync

Commit:
Merge branch 'develop' into feature/threading-gui-and-pt

henrikuuste commented on CMakeLists.txt:
```diff
@@ -10,6 +10,13 @@ include(cmake/StandardProjectSettings.cmake)
 # cuda_select_nvcc_arch_flags(ARCH_FLAGS)
 
+find_package(OpenMP)
+if (OPENMP_FOUND)
+    set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
+    set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
+    set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${OpenMP_EXE_LINKER_FLAGS}")
+endif()
```
Please change to modern CMake syntax
https://cliutils.gitlab.io/modern-cmake/chapters/packages/OpenMP.html

henrikuuste commented on src/common.h:
```diff
@@ -40,6 +42,7 @@ inline const char *mode_strings[] = {
 
 struct AppContext {
   size_t frame          = 0;
```
why have separate frame numbers? Do we care about the GUI frame?

henrikuuste commented on src/scenes/pt.cpp:
```diff
@@ -29,18 +30,20 @@ Radiance trace(Scene const &scene, Ray const &wo, TraceContext &ctx) {
   size_t changeSamples = 0;
-  for (auto &pixel : image) {
+
+#pragma omp parallel for
```
Add reduction clauses for the average and changeSamples.
https://bisqwit.iki.fi/story/howto/openmp/#TheReductionClause_2

jellosubmarine replied:
Added clauses, can't use Vec3 for avgChange, had to separate variables.

henrikuuste replied:
Just use a single float, since that is what we care a bout in the end. Move maxCoeff inside the loop and it should be close enough.

Commit:
Changed Cmake syntax. Removed frame var.
 Added reduction clauses to OMP.

Commit:
Fixed typo

henrikuuste commented on src/CMakeLists.txt:
```diff
@@ -10,6 +10,11 @@ add_executable(sfml_cuda
 # )
 # set_target_properties(sfml_cuda PROPERTIES CUDA_SEPARABLE_COMPILATION ON)
 
+find_package(OpenMP)
```
Keep this line in the main CMakeLists file, since this one is just supposed to do linking and stuff, not search for new features.

henrikuuste commented on imgui.ini:
```diff
@@ -0,0 +1,15 @@
+[Window][Debug##Default]
```
Add imgui.ini to .gitignore. This is a user specific file.

henrikuuste commented on src/common.h:
```diff
@@ -41,7 +41,6 @@ inline const char *mode_strings[] = {
 };
 
 struct AppContext {
-  size_t frame          = 0;
```
Why not just keep frame instead? or rename to spp maybe, since that is actually more descriptive?

henrikuuste commented on src/scenes/pt.cpp:
```diff
@@ -50,15 +53,19 @@ void PathTracer::render(Scene const &scene, Camera const &cam, AppContext &ctx,
       Radiance change = rSample.cwiseQuotient(radianceBuffer[idx]);
       if (not change.hasNaN()) {
-        avgChange += change;
+        avgChangeR += change[0];
```
As mentioned, just moving maxCoeff here should be fine. The calculation is not exact anyway at this point.

Commit:
PR changes
Added imgui.ini to .gitignore
Renamed ptFrame to spp
Moved findpackage(OpenMP) to main CMakeLists
Moved maxcoeff inside the loop in pt.render
